### DISCUSSION TITLE
What's coming in Unpoly 3

### DISCUSSION THREAD

**Discussion by triskweline** [unpoly/unpoly#407] [Announcements]
The next major release will be **Unpoly 3**, currently scheduled for fall of 2022.

Here are 75 slides about new features and changes: http://triskweline.de/unpoly3-slides/

[![2022-07 Unpoly 3 Overview](https://user-images.githubusercontent.com/89953/181632511-40dad490-808e-4883-9b90-ecbfcff95eac.png)](http://triskweline.de/unpoly3-slides/)

**Comment by adam12** [comment]
This looks great. A lot of the `[up-validate]` improvements are things I ran into again just recently, so extremely happy to see them coming.

Thanks! :100:

**Comment by apollo13** [comment]
The changes around caching (etags & last-modified) look great. Same goes for forms, especially with their eventual consistency.

Curious, how much will the bundle size drop with removing ie11? While unpoly isn't exactly large for todays standards, it is not really small either (I haven't yet looked into treeshaking since size was not really an issue for internal apps).

Either way, can't wait to try v3, thank you so much!

**Comment by triskweline** [comment]
> Curious, how much will the bundle size drop with removing ie11? 

It is not going to drop since we're also adding a lot of features.

> I haven't yet looked into treeshaking since size was not really an issue for internal apps

Treeshaking Unpoly is not trivial without major changes the the library. The background here is neither Unpoly nor your bundler does not know which of Unpoly's HTML attributes (`[up-...]`) or API calls in inline event handlers (`[up-on-accepted="up.apiMethod()"]`) Unpoly is going to encounter while your app is running. Hence there is no static dependency tree to shake.

**Comment by alexandremjacques** [comment]
Any ETA on version 3? Eager to try it!

**Reply by triskweline** [reply]
The code has been done for a while. It's currently being documented.The code has been completed for a while. I'm in the [process](https://github.com/unpoly/unpoly/commits/master) of documenting it.

I hope to finish the documentation by the end of November.

**Reply by ammezie** [reply]
Hi @triskweline any ETA when v3 will be released?

**Reply by triskweline** [reply]
Code and docs are done. We're currently dogfooding it on real applications to see if there are any regressions.

**Reply by ammezie** [reply]
Great!

**Reply by xinha-sh** [reply]
> Code and docs are done. We're currently dogfooding it on real applications to see if there are any regressions.

Any way I can try this on my personal project ?

**Reply by triskweline** [reply]
I released a version `3.0.0-rc2` that you can try out.

npm package:
https://www.npmjs.com/package/unpoly/v/3.0.0-rc2

CDN links:
https://unpkg.com/browse/unpoly@3.0.0-rc2/

Slide deck explaining major changes:
http://triskweline.de/unpoly3-slides/

Preliminary CHANGELOG:
https://github.com/unpoly/unpoly/blob/master/CHANGELOG.md

Upgrade guide:
https://unpoly.com/changes/upgrading

Let me know if you run into problems!

**Reply by adam12** [reply]
This is fantastic, thanks. :+1:

**Reply by apollo13** [reply]
Lovely, thanks! Are there hosted docs for v3 somewhere already?

**Reply by triskweline** [reply]
I uploaded the docs here:
https://v3.unpoly.com/

**Reply by triskweline** [reply]
I released a version `3.0.0-rc2`. I updated the links for npm/CDN in my comment above.

**Comment by apollo13** [comment]
Our projects worked basically unmodified on V3. Great work!

**Reply by triskweline** [reply]
That's useful feedback. Thanks @apollo13 !

**Reply by apollo13** [reply]
Please note that our applications are rather simple, so it is probably not the best indicator. Nevertheless it was a smooth upgrade :)

**Comment by EugeneDae** [comment]
@triskweline Hats off to the team behind Unpoly and you personally 👏🏻 You guys are absolute legends not just for creating the thing, but also sharing it with the world, putting effort into popularizing it (having the courage to go against the grain of the industry) and, of course, maintaining it for years.

I'm new to Unpoly, but I'm here to stay and can't wait to learn it in full to use in my Django projects.

As someone who's spent so much time dealing with ever increasing complexity of JS-based SPA stacks, my only regret is that I didn't start learning it earlier.

I hope more devs around the world try Unpoly and unleash their productivity with it.

Bravo!

**Reply by triskweline** [reply]
@EugeneDae Thank you so much for your very kind words :heart:

**Comment by triskweline** [comment]
I released a version `3.0.0-rc4` that you can try out:

> npm package:
> <https://www.npmjs.com/package/unpoly/v/3.0.0-rc4>
> 
> CDN links:
> <https://unpkg.com/browse/unpoly@3.0.0-rc4/>
> 
> Slide deck explaining major changes:
> <http://triskweline.de/unpoly3-slides/>
>  
> Documentation:
> <https://v3.unpoly.com>
> 
> Preliminary CHANGELOG:
> <https://v3.unpoly.com/changes/3.0.0-rc4>
> 
> Upgrade guide:
> <https://v3.unpoly.com/changes/upgrading>

The biggest change since `3.0.0-rc2` is a new cache implementation:

> Unpoly has always allowed server-side code to inspect [request headers](/up.protocol) to [customize or shorten responses](/optimizing-responses), e.g. by omitting content that isn't [targeted](/targeting-fragments). Unpoly makes some changes how optimized responses are [cached](/caching):
> 
> - Requests with the same URL and HTTP method, but different header values (e.g. `X-Up-Target`), now share the same cache entry.
> - If a server optimizes its response, all request headers that influenced the response should be listed in a `Vary` response header. This tells Unpoly to partition its [cache](/caching) for that URL so that each request header value gets a separate cache entries.
> - Sending `Vary` headers also prevents browsers from using an optimized response for full page loads.
> - The configuration `up.network.config.requestMetaKeys` has been removed.

You can contribute by testing this release candidate against your existing Unpoly apps.

If no further issues are discovered, this will be released as 3.0.0 in the next few days.

**Reply by naveencom** [reply]
Thanks a lot for the hardwork!

**Reply by simosentissi** [reply]
Always amazed and grateful at how much work y'all are putting into unpoly!

**Comment by szabeszg** [comment]
@triskweline First of all, thanks for all the hard work that you and your team puts into sharing Unpoly.

Second, this new behaviour is very problematic in my case:
_"After rendering stale content from the cache, Unpoly now automatically renders a second time with fresh content from the server."_

I have a cart page which now first displays stale content and after the HXR request it gets updated but this is huge issue I think, because what if the XHR update takes seconds, for example? Even with a 200-300 ms waiting time it is rather strange to see stale Cart Items come and go (depending on if the item was added or removed from the cart).

Note that I use POST request to change the cart so this should apply:
_"When the user makes a non-GET request (usually a form submission with POST), the entire cache is expired."_

Sending X-Up-Evict-Cache: /cart/* from the server partially solves the issue, because in the case of deleting an item from the cart and revisiting the cart again no stale content is shown with this header sent. But it does not solve the action of adding an item when the user sees it being added to the cart only after the XHR request succeeds.

What can be done to see fresh content right away all the time (no matter what) for a give page? Am I missing something perhaps?

**Reply by triskweline** [reply]
Hi @szabeszg ! Thanks for your feedback.

> Sending X-Up-Evict-Cache: /cart/* from the server partially solves the issue, because in the case of deleting an item from the cart and revisiting the cart again no stale content is shown with this header sent. But it does not solve the action of adding an item when the user sees it being added to the cart only after the XHR request succeeds.

You can send the same header from the server route that adds an item.

> What can be done to see fresh content right away all the time (no matter what) for a give page? 

You can [disable caching](https://v3.unpoly.com/caching#disabling-caching) globally, per link, or for all requests to a given URL.

Or, if you prefer Unpoly 2's behavior of evicting the entire cache after a request with an [unsafe](https://developer.mozilla.org/en-US/docs/Glossary/Safe/HTTP) HTTP method, configure the following:

```js
up.network.config.evictCache = (request) => !request.isSafe()
```

**Reply by szabeszg** [reply]
@triskweline Thank you for your kind and quick response!
> ```js
> up.network.config.evictCache = (request) => !request.isSafe()
> ```
As a quick solution, I will use this config option to get back the old behaviour, thanks for making me aware of it. I applied it and it works as expected.

As soon as I have the time, I will look into how the new behaviour can be used so that it does not introduce this "updating in front of the eyes of the user" _on pages where it is puzzling to see it_.

**Comment by alnuaimi94** [comment]
Why don't you split the unpoly library into small packages that are downloaded via npm when you need it This gives less library size and improves site performance You have for example swup js library that does that
